docs/tutorial/programming.md

```diff
@@ -1,10 +1,295 @@
+| base addr 2             | 9                        |   x3c9         | 
+| start                   | 10                       |   x3ca         | 
+| streamer perf. counter  | 11                       |   x3cb         |  
+| mode                    | 12                       |   x3cc         |
+| length                  | 13                       |   x3cd         |
+| start                   | 14                       |   x3ce         |
+| busy                    | 15                       |   x3cf         |
+| alu perf. counter       | 16                       |   x3d0         |
+
+!!! note
+
+    In our SNAX system, we reserved the CSR address spaces from 0x3c0 up to 0x5ff.
+
+In the `snax-alu.c` file you can see that we can write the CSR registers directly. The `write_csr` and `read_csr` functions call inline assembly calls for `csrw` and `csrr` RISCV CSR instructions, respectively:
+
+```C
+//------------------------------
+// 1st set the streamer CSRs
+// The list of CSRs are:
+// 0x3c0 - loop bound for all components (RW)
+// 0x3c1 - temporal stride for input A (RW)
+// 0x3c2 - temporal stride for input B (RW)
+// 0x3c3 - temporal stride for output O (RW)
+// 0x3c4 - spatial stride for input A (RW)
+// 0x3c5 - spatial stride for input B (RW)
+// 0x3c6 - spatial stride for input O (RW)
+// 0x3c7 - base pointer for input A (RW)
+// 0x3c8 - base pointer for input B (RW)
+// 0x3c9 - base pointer for input O (RW)
+// 0x3ca - send configurations to streamer (RW)
+// 0x3cb - performance counter of streamer (RO)
+//------------------------------
+write_csr(0x3c0, LOOP_ITER);
+write_csr(0x3c1, 32);
+write_csr(0x3c2, 32);
+write_csr(0x3c3, 64);
+write_csr(0x3c4, 8);
+write_csr(0x3c5, 8);
+write_csr(0x3c6, 8);
+write_csr(0x3c7, (uint64_t)local_a);
+write_csr(0x3c8, (uint64_t)local_b);
+write_csr(0x3c9, (uint64_t)local_o);
+write_csr(0x3ca, 1);
+```
+
+The sequence above first configures the streamers and then starts the streamers. Recall that our streamer supports the following loop for accessing data:
+
+```
+for(i = 0; i < N; i++):
+  # Parfor equivalent
+  parfor(j = 0; j < 4; j++):
+    target_address[j] = base_address + temporal_stride*i + spatial_stride*j;
+```
+Where `N` is the configured loop-bound register. The `base_address` for each port `A`, `B`, and `OUT` are configured with `(uint64_t)local_a`, `(uint64_t)local_b`, and `(uint64_t)local_o`, respectively.
+
+The temporal strides for each input are 32 bytes, but the output is 64 bytes. These addresses are in bytes. We need 32 for the inputs because that's 8 bytes (64 bits) per PE and we have 4 PEs and therefore, 32 bytes of temporal stride. The output is 64 because we need 16 bytes (128 bits) per PE and we have 4 PEs and therefore, 64 bytes of temporal stride.
+
+All spatial strides are 8 because, for each TCDM port connected to the streamer, it will automatically increment or point to the next bank which is in byte addresses. 
```

Comment:
```suggestion
All spatial strides are 8 because, for each TCDM port connected to the streamer, it will automatically increment or point to the next bank which is 8 in byte addresses (each bank is 64bit, aka, 8 bytes). 
```